Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/checkins/checkout/ContactJournalCheckInEntryCreator.kt

```diff
@@ -0,0 +1,104 @@
+package de.rki.coronawarnapp.presencetracing.checkins.checkout
+
+import androidx.annotation.VisibleForTesting
+import dagger.Reusable
+import de.rki.coronawarnapp.contactdiary.model.ContactDiaryLocation
+import de.rki.coronawarnapp.contactdiary.model.ContactDiaryLocationVisit
+import de.rki.coronawarnapp.contactdiary.model.DefaultContactDiaryLocation
+import de.rki.coronawarnapp.contactdiary.model.DefaultContactDiaryLocationVisit
+import de.rki.coronawarnapp.contactdiary.storage.repo.ContactDiaryRepository
+import de.rki.coronawarnapp.eventregistration.checkins.CheckIn
+import de.rki.coronawarnapp.eventregistration.checkins.split.splitByMidnightUTC
+import de.rki.coronawarnapp.util.TimeAndDateExtensions.toLocalDateUtc
+import de.rki.coronawarnapp.util.TimeAndDateExtensions.toUserTimeZone
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.firstOrNull
+import org.joda.time.Duration
+import org.joda.time.Seconds
+import org.joda.time.format.DateTimeFormat
+import timber.log.Timber
+import javax.inject.Inject
+import kotlin.math.roundToLong
+
+@Reusable
+class ContactJournalCheckInEntryCreator @Inject constructor(
+    private val diaryRepository: ContactDiaryRepository
+) {
+
+    suspend fun createEntry(checkIn: CheckIn) {
+        Timber.d("Creating journal entry for %s", checkIn)
+
+        // 1. Create location if missing
+        val location: ContactDiaryLocation = diaryRepository.locations.first()
+            .find { it.traceLocationID == checkIn.traceLocationId } ?: checkIn.toLocation()
+
+        // 2. Split CheckIn by Midnight UTC
+        val splitCheckIns = checkIn.splitByMidnightUTC()
+        Timber.d("Split %s into %s ", this, splitCheckIns)
+
+        // 3. Create LocationVisit if missing
+        splitCheckIns
+            .createMissingLocationVisits(location)
+            .forEach { diaryRepository.addLocationVisit(it) }
+    }
+
+    private suspend fun CheckIn.toLocation(): ContactDiaryLocation {
+        val location = DefaultContactDiaryLocation(
+            locationName = locationName(),
+            traceLocationID = traceLocationId
+        )
+        Timber.d("Created new location %s and adding it to contact journal db", location)
+        return diaryRepository.addLocation(location) // Get location from db cause we need the id autogenerated by db
+    }
```

Comment:
Please use a different name or pattern here. `toLocation` suggests conversions (compare all other cases of `toABC`), but this actually performs a database write operation, :ghost:.

Reply:
:ghostbusters

Reply:
In this case it won't make a difference. If it's null the test will fail anyways